Source/Services.Clients/ReverseCallClient.cs

```diff
@@ -244,121 +159,293 @@ protected virtual void Dispose(bool disposing)
+                catch (Exception exception)
                 {
-                    _writeResponseSemaphore.Release();
+                    _metrics.IncrementTotalFailedRequestCallbacks();
+                    _logger.HandlingRequestFailedToInvokeCallback(typeof(TRequest), callId, exception);
                 }
             }
-            catch (Exception ex)
+            catch (Exception exception)
             {
-                _logger.LogWarning(ex, "An error occurred while handling received request");
+                _metrics.IncrementTotalFailedRequestHandlers();
+                _logger.HandlingRequestFailed(typeof(TRequest), exception);
             }
         }
 
-        Task WriteResponse(TResponse response, ReverseCallId callId, CancellationToken cancellationToken)
+        async Task WriteResponse(TResponse response, ReverseCallId callId, CancellationToken cancellationToken)
         {
-            var responseContext = new ReverseCallResponseContext { CallId = callId.ToProtobuf() };
-            _setResponseContext(response, responseContext);
-            var message = new TClientMessage();
-            _setMessageResponse(message, response);
-            if (!cancellationToken.IsCancellationRequested)
+            try
+            {
+                var responseContext = new ReverseCallResponseContext { CallId = callId.ToProtobuf() };
+                _protocol.SetResponseContext(responseContext, response);
+                var message = new TClientMessage();
+                _protocol.SetResponse(response, message);
+
+                await WriteMessage(message, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception exception)
+            {
+                _metrics.IncrementTotalFailedResponseWrites();
+                _logger.HandlingRequestFailedToWriteResponse(typeof(TRequest), callId, exception);
+            }
+        }
+
+        async Task WriteMessage(TClientMessage message, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            _logger.WritingMessage(typeof(TClientMessage));
+
+            if (!_writeLock.Wait(0))
             {
-                _logger.WritingResponse(callId);
-                return _clientToServer.WriteAsync(message);
+                var stopwatch = Stopwatch.StartNew();
+
+                _metrics.IncrementPendingWrites();
+                _logger.WritingMessageBlockedByAnotherWrite(typeof(TClientMessage));
+
+                await _writeLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+                try
+                {
+                    stopwatch.Stop();
+                    _metrics.DecrementPendingWrites();
+                    _metrics.AddToTotalWriteWaitTime(stopwatch.Elapsed);
+                    _logger.WritingMessageUnblockedAfter(typeof(TServerMessage), stopwatch.Elapsed);
+                }
+                catch
+                {
+                }
```

Comment:
This is a bit of a weird try catch, could it just be taken out? Or it could be changed to cover the whole pending block and then we could have logs/metrics for `FailedPendingWrite` or something.
```suggestion

                stopwatch.Stop();
                _metrics.DecrementPendingWrites();
                _metrics.AddToTotalWriteWaitTime(stopwatch.Elapsed);
                _logger.WritingMessageUnblockedAfter(typeof(TServerMessage), stopwatch.Elapsed);
```

Reply:
I agree it looks a bit strange - but it is necessary to not leave locks locked if an error occurs. We can look into a nicer flow in the future (we have the same thing in WrappedAsyncStreamWriter).